### DISCUSSION TITLE
Question about softmax

### DISCUSSION THREAD

**Discussion by morgan-bc** [pyg-team/pytorch_geometric#3485] [Q&A] [answered]
In the softmax in PyG is 

```python
N = maybe_num_nodes(index, num_nodes)
src_max = scatter(src, index, dim, dim_size=N, reduce='max')
src_max = src_max.index_select(dim, index)
out = (src - src_max).exp()
out_sum = scatter(out, index, dim, dim_size=N, reduce='sum')
out_sum = out_sum.index_select(dim, index)

out = out / (out_sum + 1e-16)
```

The equation corresponding the code is
![pyg_01](https://user-images.githubusercontent.com/15273348/141418306-c7cb138c-08bd-428d-863a-2603f3961db4.JPG)

The original softmax equation is
![pyg_02](https://user-images.githubusercontent.com/15273348/141418371-9d81a1b9-ec5f-408c-b01f-285618630907.JPG)

Why dose the PyG softmax  subtract the max value of the neighborhood nodes ?

**Comment by wsad1** [comment, marked as answer]
Both the equations you pasted will yield the same output, as softmax(x) = softmax(x+c) where c is some constant.
Now setting c to max(x) makes the softmax more stable by preventing underflow and overflow, read more [here](https://stackoverflow.com/questions/42599498/numercially-stable-softmax).
